### ISSUE TITLE
Add SASS support

### ISSUE THREAD

**Issue by bep** [gohugoio/hugo#4243]
This is a discussion. Thinking out loud.

With the Chroma for highlighting and the image processing added in `0.32` you have an almost complete super-fast web dev environment.

But there is one other common task that forces people to use Gulp and friends (there are of course other requirements, too, but this is the most common); to compile your CSS. With Bootstrap 4 I guess the most common one is SASS.

There is no SASS compiler in native Go, and I suspect that it will take a decade before that happens.
But there is one in C/C++ with many language wrappers:

https://github.com/wellington/go-libsass

**This will, sadly, complicate the build setup. But it is doable, the question is: Is it worth it?**

I'm not envisioning any "asset pipeline" for this ... Maybe we could just add `SASS` as a new resource type:

```html
{{ (((.Resources.ByType "sass" ).GetByPrefix "main").ConvertTo "css").RelPermalink }} 
```

Or something.

Thoughts? Maybe this should wait until we "land" the plugin debate.

/cc @spf13

**Comment by RealOrangeOne** [comment]
Assuming the resources were able to sit in the `static/` rather than in `content/`. This sounds like a great idea. 

I can't see this adding any additional _runtime_ dependencies (Go should statically link the `libsass` library at compile time?). Build dependencies arent a massive issue, as installing libraries like `libsass` is often very easy. It might also be possible to only require `libsass` to be linked if youre calling `.ConvertTo "css" (very possible I believe).

Agreed this should be considered after plugins, as if plugins are decided as a good course of action, I think something this shouldn't be in Hugo core.

`</TwoCents>`

**Comment by Jos512** [comment]
Without an asset pipeline, I doubt adding SASS support is worth the work. That is, I think that most people using SASS also perform other front-end tooling tasks. So they can port one task to Hugo (SASS), but use their toolchain for the others.

(This is not me saying it's a bad idea. I just respond to the question of 'is it worth it?', which from my personal perspective I doubt. Is perhaps a poll on the Discourse forum an option to estimate the interest? I'm of course responding from my biased perspective.)

**Comment by RealOrangeOne** [comment]
@Jos512 I hadn't thought about other build lines (even though I use them myself -_- ). I think the plugin interface would solve this though? Plugins could then be created to support the different files.

**Comment by rdwatters** [comment]
I don’t think it’s worth it since SASS is most commonly one step in frontend builds along with JS. 

I like the clean separation between Hugo with *content* (HTML, MD, JSON, images, any other text File, etc) and users picking whatever toolchain they want for *presentation* and *behavior*. Most Jekyll users I have talked to (Jekyll ships with SASS and Coffeescript support, I believe) don’t use the built-in asset pipeline, but this isn’t a data-based comment and might have to do with the Ruby implementations being slow. 

I think the JS world here (Webpack CS vs Gulp, PostCSS vs SASS, etc) is really difficult to keep up with. 

cc/ @budparr

**Comment by biodranik** [comment]
1. Supporting SASS is a good idea, almost everyone uses it.
2. Is go-libsass well supported? SASS is actively developing, and it would be sad to stick with an outdated implementation.
3. Plugins are the way to go. And it should be very easy in the basic implementation: just provide an easy way for authors to call any pre- and post-processing external tools they want. And disable it by-default for themes for security.
4. Some of the tools should be run on a file change (e.g. SASS, autoprefixer, etc.), and it would be great to expose `hugo server` watch functionality to external tools, by simply binding a custom command line to modifications of specific files. This solution is not a fully flexible plugin, but it solves the most painful issues during development. Build-time functionality can be solved now by a custom shell/cmd script.

**Comment by devjack** [comment]
Hugo is, for me, a fantastic generation tool that plays an important part in my publishing pipeline. I use it in combination with webpack/gulp/sass etc. to produce sites. My gut tells me documentation/examples/ecosystem around combined tooling is more beneficial than replicating existing community functionality in hugo.

**Comment by fbnlsr** [comment]
SASS support would be a great addition to Hugo. I agree with @devjack though, focus should be on documentation and ecosystem, as it's rather simple to set up a Webpack / Gulp tooling when working with Hugo. Especially when considering the first post stating that even if it's doable, it's something that would be hard to implement and complicate the build setup.

One thing that could be more beneficial though is a hook system to trigger commands before and after building. At the moment I'm using NPM tasks to run my build/serve/watch tasks, which look like this:

```json
"scripts": {
    "build": "webpack && hugo",
    "serve": "webpack && hugo serve",
    "watch": "parallelshell 'hugo serve' 'webpack --watch'"
  }
```
A simple built-in hook system in Hugo could help users set up a more streamlined development environment and be more flexible imho. It'd also help reduce the number of dependencies needed.

**Comment by biodranik** [comment]
It would be great to reuse hugo's *watch* functionality and add custom hooks on file changes for more flexibility. In this case it would be very easy to call any external script/tool/sass/webpack/anything to preprocess files.

There is no any easy/cross-platform/no-filesystem-polling solution available, and golang portability can solve it.

**Comment by andi1984** [comment]
@bep I'm with @devjack opinion: static assets like images, audio, video etc. is in general fine for me when Hugo optimizes them. But I think it is not worth effort trying to build something others already do. I'm using Sass and Gulp myself. But might be that in few months it will be grunt, postcss or whatever else is out there.

I think Hugo should not care about that.

**Comment by nfisher** [comment]
@bep I think I could probably add some rudimentary SASS capabilities with PR #4278.

`{{ sass OUTFILE, INFILE }}`

The main bit I need to figure out is the best way to hook into memory server/http handler. Which I need to do anyway for that PR.

**Comment by comaldave** [comment]
### I am placing my TS and SCSS files in /src/ and then process them along with webpack to create CSS and JS in /static/

@Biodranik
> It would be great to reuse hugo's watch functionality and add custom hooks on file changes for more flexibility. In this case it would be very easy to call any external script/tool/sass/webpack/anything to preprocess files.

**Comment by fbnlsr** [comment]
Same here. I ended up using Webpack as well. I've got an npm script which launches webpack and hugo in parallel with the watch flag and it's working like a charm.

You can see it here: https://github.com/fbnlsr/primative.net

**Comment by xzyfer** [comment]
LibSass maintainer here 👋 happy to help out or answer any questions

**Comment by jhabdas** [comment]
@xzyfer I noticed Discourse is using Sass. Do you know offhand if they're using LibSass or Rust?

**Comment by jhabdas** [comment]
Thanks for the info. And I apologize for getting my four-letter languages mixed up. I actually meant DART and not Rust. The reason I asked is because the [Ruby implementation](https://github.com/sass/ruby-sass) is now deprecated (tentatively) and I see the [Dart implementation ](https://github.com/sass/dart-sass) picking up a lot of steam.

I'm actually quite thrilled to see the move to Dart as it may breathe new life into Sass and, as a result, lead Discourse to start porting away from the Ruby implementation to either LibSass or DartSass.

**Comment by jhabdas** [comment]
> I think the JS world here (Webpack CS vs Gulp, PostCSS vs SASS, etc) is really difficult to keep up with.

@rdwatters I want to play Devil's advocate and just suggest one of the best parts of Hugo is the opinion which comes along with it. Sass may not have the newest, shiniest stuff but it's been battle tested by 10,000 angry Ruby developers, is also the language being used in Discourse 2 and intuition tells me it's going to see renewed interest soon as a result of a) the Dart implementation which is seeing some heavy development right now and b) [JavaScript developer frustration](https://medium.com/@jhabdas/webpack-is-your-achilles-heel-d3cd80821a4f).

That said, there is such a thing as too much opinion but having the ability to use a preprocessor like Sass to Hugo would add quite a bit of power and allow Hugo to slurp up a lot more Jekyll sites.

**Comment by bep** [comment]
@jhabdas is right; this is an opinionated choice, but it's very useful to a bunch of people (me included), and even if is opinionated it is also really the only choice we could make if we wanted to add a fast CSS preprocessor in Hugo.

I have said it before, but I will say this again: I'm willing to give away lots of flexibility and modern buzz word components in my workflow I that means that I can use `hugo` for all of it.

**Comment by jhabdas** [comment]
sorry couldn't help it ^^
<img width="263" alt="screen shot 2018-06-22 at 11 04 32 pm" src="https://user-images.githubusercontent.com/440298/41783928-b8ea976c-7670-11e8-8690-8f9d9d8750e6.png">

this message will self-destruct in 30 seconds…

**Comment by bep** [comment]
@xzyfer are these WARNINGs familiar?

```bash
# github.com/gohugoio/hugo/vendor/github.com/wellington/go-libsass/libs
In file included from cencode.c:2:
../vendor/github.com/wellington/go-libsass/libs/../libsass-build/cencode.c:50:5: warning: declaration does not declare anything [-Wmissing-declarations]
../vendor/github.com/wellington/go-libsass/libs/../libsass-build/cencode.c:64:5: warning: declaration does not declare anything [-Wmissing-declarations]
```

I assume "no harm", but they are annoying to get every time I build. Know any magic flag I can set to turn these warnings off?

**Comment by xzyfer** [comment]
I don't but I can take a look. A bunch of compiler warning were introduced in a recent major refactor I wasn't involved in. Safe to ignore these for now but we'll continue to address these as time permits.

**Comment by bep** [comment]
FIxed in dea71670c059ab4d5a42bd22503f18c087dd22d4

**Comment by xzyfer** [comment]
:tada:
